### ISSUE TITLE
Missing first Esc keypress after a window is focused

### ISSUE THREAD

**Issue by jneem** [pop-os/cosmic-comp#680]
When I focus a  window, the first time I press Esc I get a key release event but not a key press event. After that, each time I press Esc I get both a press and a release. Until I switch focus away and back again, and then once again the first time I press Esc I only get the release.

- It happens both in tiled mode and floating mode
- It happens whether I use the keyboard or the mouse to change focus
- I've tested several different keys, and only Esc seems to have this behavior
- I *think* this is cosmic-comp's fault because it doesn't happen under sway

**Comment by Drakulix** [comment]
Which commit is this on? Can't reproduce with latest master.

**Comment by jneem** [comment]
Thanks for looking into it! It's `2cef99ae783efb451bd9067b4d9e74c014615996`, as packaged and configured by [nixos-cosmic](https://github.com/lilyinstarlight/nixos-cosmic). I'll try to see if I can reproduce it with "vanilla" cosmic-comp.

**Comment by Drakulix** [comment]
Even on that commit I don't see any missing events. For testing I have been running clients with both `WAYLAND_DEBUG=1` and `xtrace` and for wayland clients I get the correct `wl_keyboard` events as well for Xwayland clients the corresponding `KeyPress` and `KeyRelease` events.

Any particular application / test setup you are using?

**Comment by jneem** [comment]
I first ran into this "in the wild" using alacritty and helix. I tested it by modifying the `winit` window example to print out the key events it received. I just tried running that example with `WAYLAND_DEBUG=1`, and this is what I get:

```
[4171920.646] wl_keyboard@21.enter(1347465, wl_surface@29, array[4])
[4171920.687] zwp_text_input_v3@22.enter(wl_surface@29)
[4171920.697] wl_keyboard@21.modifiers(1347465, 64, 0, 0, 0)
[4171920.705] xdg_toplevel@31.configure(1917, 1025, array[20])
[4171920.713] xdg_surface@30.configure(1347467)
[4171920.782]  -> zwp_text_input_v3@22.enable()
[4171920.794]  -> zwp_text_input_v3@22.set_content_type(0, 0)
[4171920.800]  -> zwp_text_input_v3@22.commit()
[4171920.837]  -> xdg_surface@30.ack_configure(1347467)
2024-08-06T01:47:07.874840Z  INFO window: Window=WindowId(94358172653328) focused
2024-08-06T01:47:07.874888Z  INFO window: IME enabled for Window=WindowId(94358172653328)
2024-08-06T01:47:07.874895Z  INFO window: Modifiers changed to ModifiersState(SUPER)
[4171920.992] wl_buffer@39.release()
[4171934.724]  -> wl_surface@29.frame(new id wl_callback@36)
[4171934.756]  -> wl_surface@29.attach(wl_buffer@39, 0, 0)
[4171934.764]  -> wl_surface@29.damage_buffer(0, 0, 1917, 1025)
[4171934.768]  -> wl_surface@29.commit()
[4171934.831] wl_keyboard@21.modifiers(1347468, 0, 0, 0, 0)
[4171934.841] wl_keyboard@21.modifiers(1347471, 64, 0, 0, 0)
2024-08-06T01:47:07.888812Z  INFO window: Modifiers changed to ModifiersState(0x0)
2024-08-06T01:47:07.888832Z  INFO window: Modifiers changed to ModifiersState(SUPER)
[4171946.700] wl_display@1.delete_id(36)
[4171946.716] wl_callback@36.done(86934126)
[4172029.915] wl_keyboard@21.key(1347473, 86934208, 36, 0)
2024-08-06T01:47:07.984018Z  INFO window: key Character("j"), state Released
[4172244.260] wl_keyboard@21.key(1347475, 86934422, 125, 0)
[4172244.291] wl_keyboard@21.modifiers(1347476, 0, 0, 0, 0)
2024-08-06T01:47:08.198315Z  INFO window: key Named(Super), state Released
2024-08-06T01:47:08.198403Z  INFO window: Modifiers changed to ModifiersState(0x0)
[4172854.267] wl_keyboard@21.keymap(1, fd 11, 64754)
[4172854.328] wl_keyboard@21.modifiers(1347479, 0, 0, 0, 0)
[4172854.339] wl_keyboard@21.key(1347478, 86935029, 1, 0)
2024-08-06T01:47:08.810467Z  INFO window: Modifiers changed to ModifiersState(0x0)
2024-08-06T01:47:08.810499Z  INFO window: key Named(Escape), state Released
[4173193.542] wl_keyboard@21.key(1347480, 86935372, 1, 1)
2024-08-06T01:47:09.147689Z  INFO window: key Named(Escape), state Pressed
[4173193.896] wl_keyboard@21.key(1347481, 86935373, 1, 0)
2024-08-06T01:47:09.147914Z  INFO window: key Named(Escape), state Released
```

This is me focusing the window (using Super+j, so that explains the initial modifier state and the first 2 key releases) and then pressing Esc twice. The first press only sends the release; the second one sends both.

**Comment by jneem** [comment]
Ok, I just noticed something new: this only happens when I'm running fcitx. With fcitx turned off, I get
```
[ 771672.503] wl_keyboard@21.enter(8994, wl_surface@29, array[4])
[ 771672.547] wl_keyboard@21.modifiers(8994, 64, 0, 0, 0)
[ 771672.566] xdg_toplevel@31.configure(1277, 1528, array[20])
[ 771672.575] xdg_surface@30.configure(8996)
[ 771672.687]  -> xdg_surface@30.ack_configure(8996)
2024-08-06T02:02:02.593986Z  INFO window: Window=WindowId(94224799999760) focused
2024-08-06T02:02:02.594013Z  INFO window: Modifiers changed to ModifiersState(SUPER)
[ 771672.813] wl_buffer@39.release()
[ 771687.008]  -> wl_surface@29.frame(new id wl_callback@36)
[ 771687.032]  -> wl_surface@29.attach(wl_buffer@39, 0, 0)
[ 771687.039]  -> wl_surface@29.damage_buffer(0, 0, 1277, 1528)
[ 771687.044]  -> wl_surface@29.commit()
[ 771691.886] wl_display@1.delete_id(36)
[ 771691.901] wl_callback@36.done(87828838)
[ 771833.803] wl_keyboard@21.key(8997, 87828980, 38, 0)
2024-08-06T02:02:02.755381Z  INFO window: key Character("l"), state Released
[ 771972.809] wl_keyboard@21.key(8998, 87829119, 125, 0)
[ 771972.861] wl_keyboard@21.modifiers(8998, 0, 0, 0, 0)
2024-08-06T02:02:02.894412Z  INFO window: key Named(Super), state Released
2024-08-06T02:02:02.894463Z  INFO window: Modifiers changed to ModifiersState(0x0)
[ 772962.823] wl_keyboard@21.key(8999, 87830109, 1, 1)
2024-08-06T02:02:03.884360Z  INFO window: key Named(Escape), state Pressed
[ 772963.288] wl_keyboard@21.key(9000, 87830109, 1, 0)
2024-08-06T02:02:03.884615Z  INFO window: key Named(Escape), state Released
```

I did just go and check sway again, and I get the Esc press event whether or not fcitx is running

**Comment by jneem** [comment]
After a long hiatus, I've started digging into this again. Some print-debugging reveals that the escape keypress is being filtered out in `State::filter_keyboard_input`, because the keyboard is always grabbed for the first `Escape` keypress on a newly-focused window. The grab is always an `InputMethodKeyboardGrab`.

The only place that I see where this can come from is from a `zwp_input_method_v2::Request::GrabKeyboard` request, but I've tried running `WAYLAND_DEBUG=1` for both the window that I'm switching to and the window that I'm switching from and I don't see any `zwp_input_method_v2` events.

I see that the grab is unset when the `ZwpInputMethodKeyboardGrabV2` is `destroy()`ed, but not when it [gets a Release request](https://github.com/Smithay/smithay/blob/14c7e43255968e2d979a19e1fd9bd694e5a9726e/src/wayland/input_method/input_method_keyboard_grab.rs#L127). Maybe the grab should be unset there also?

**Comment by chkoreff** [comment]
I recently upgraded to Pop OS 24 (COSMIC) and I'm noticing that very often I have to press ESC twice to return to normal mode in vim.  This is annoying.  It didn't happen before COSMIC.  I'm running vim 9.1.697.  I have just begun to research this issue.  Does anyone have a current resolution for this bug?

**Comment by pmconne** [comment]
I am having the same issue as @chkoreff using nvim v0.12.0-dev. Repro steps:
1. Have one window open (e.g., Firefox).
1. Open COSMIC terminal in a second window and start nvim.
1. In nvim, press `iabc` followed by `Esc`. This inserts the text "abc" and returns to normal mode.
1. Mouse-click on the first open window, then on COSMIC terminal.
1. In nvim, press `iabc` followed by `Esc`.
  - Expected result: This should insert the text "abc" and return to normal mode.
  - Actual result: It inserts the text and remains in insert mode.
1. Press `Esc` again. Now we are back in normal mode.

In other words, if the terminal loses and re-acquires focus as a result of mouse clicks, the first `Esc` keypress upon refocus has no effect.
If I use alt-tab instead of mouse clicks, the issue is not reproducible.

The same behavior can be observed when using `vi` mode in the terminal (`set -o vi`).

I like COSMIC terminal but this is a deal-breaker for me.

**Comment by chkoreff** [comment]
> I am having the same issue as [@chkoreff](https://github.com/chkoreff) using nvim v0.12.0-dev. Repro steps:
...
> I like COSMIC terminal but this is a deal-breaker for me.

For that very reason I am using the XFCE terminal.
